pkg/blockstorage/zone/zone.go

```diff
@@ -39,153 +36,89 @@ type (
 )
 
 // FromSourceRegionZone gets the zones from the given region and sourceZones
-// It will return a minimum of 0 and a maximum of zones equal to the length of sourceZones.
+// It will return a minimum of 0 and a maximum of zones equal to the length zones available from kubernetes.
 // Depending on the length of the slice returned, the blockstorage providers will decide if
 // a regional volume or a zonal volume should be created.
-func FromSourceRegionZone(ctx context.Context, m Mapper, region string, sourceZones ...string) ([]string, error) {
+func FromSourceRegionZone(ctx context.Context, m Mapper, kubeCli kubernetes.Interface, sourceRegion string, sourceZones ...string) ([]string, error) {
 	newZones := make(map[string]struct{})
-	cli, err := kube.NewClient()
-	if err == nil {
-		nzs, rs, errzr := NodeZonesAndRegion(ctx, cli)
-		if errzr != nil {
-			log.WithError(errzr).Print("Ignoring error getting Node availability zones.")
-		}
-		if len(nzs) != 0 {
-			for _, zone := range sourceZones {
-				var z string
-				// getZoneFromKnownNodeZones() func makes sure that all zones
-				// added to newZones[] are unique and not repeated.
-				z, err = getZoneFromKnownNodeZones(ctx, zone, nzs, newZones)
-				if err == nil && isZoneValid(ctx, m, z, rs) {
-					newZones[z] = struct{}{}
-				}
-				if err != nil {
-					log.WithError(err).Print("Ignoring error getting Zone from KnownNodeZones.")
-				}
-			}
-		}
+	validZoneNames, err := m.FromRegion(ctx, sourceRegion)
+	if err != nil || len(validZoneNames) == 0 {
+		return nil, errors.Wrapf(err, "No provider zones for region (%s)", sourceRegion)
 	}
+	if !isNil(kubeCli) {
+		getAvailableZones(ctx, newZones, kubeCli, validZoneNames, sourceZones, sourceRegion)
+	}
+	// If Kubernetes provided zones are invalid use valid sourceZones
 	if len(newZones) == 0 {
+		log.Info().Print("Validating source zones")
 		for _, zone := range sourceZones {
-			// WithUnknownNodeZones() func makes sure that all zones
-			// added to newZones[] are unique and not repeated.
-			newZone := WithUnknownNodeZones(ctx, m, region, zone, newZones)
-			if newZone != "" {
-				newZones[newZone] = struct{}{}
+			if isZoneValid(zone, validZoneNames) {
+				newZones[zone] = struct{}{}
 			}
 		}
 	}
-
 	if len(newZones) == 0 {
-		return nil, errors.Errorf("Could not get zone for region %s and sourceZones %s", region, sourceZones)
+		return nil, errors.Errorf("Unable to find valid availabilty zones for region (%s)", sourceRegion)
 	}
-
 	var zones []string
 	for z := range newZones {
 		zones = append(zones, z)
 	}
 	return zones, nil
 }
 
-func isZoneValid(ctx context.Context, m Mapper, zone, region string) bool {
-	if validZones, err := m.FromRegion(ctx, region); err == nil {
-		for _, z := range validZones {
-			if zone == z {
-				return true
-			}
-		}
-	}
-	return false
+func isNil(i interface{}) bool {
+	return i == nil || reflect.ValueOf(i).IsNil()
```

Comment:
reflection can be expensive. Why use this over a typed nil comparison?

Reply:
Sure, seemed easier but will switch over to nil comparison